libs/contrib/Decidable/Decidable/Extra.idr

```diff
@@ -0,0 +1,25 @@
+module Decidable.Decidable.Extra
+
+import Data.Rel
+import Data.Fun
+import Data.Vect
+import Data.Fun.Extra
+import Decidable.Decidable
+
+public export
+NotNot : {ts : Vect n Type} -> (r : Rel ts) -> Rel ts
+NotNot r = map @{Nary} (Not . Not) r 
+
+[DecidablePartialApplication] {x : t} -> (tts : Decidable (t :: ts) r) => Decidable ts (r x) where
+  decide = decide @{tts} x
+
+public export
+doubleNegation : {ts : Vect n Type} -> {r : Rel ts} -> Decidable ts r =>
```

Comment:
We should probably use a different name because we ought to
be able to prove something similar for `All ts (NotNot r)`.

Reply:
Good point! Thanks!
I've now refactored it into: 
```idris
doubleNegationElimination : ... (witness : HVect ts) -> uncurry (NotNot {ts} r) witness -> uncurry              r  witness
doubleNegationForall : ... All ts (NotNot {ts} r) -> All ts r
doubleNegationExists : ...Ex ts (NotNot {ts} r) -> Ex ts r
```
and added introduction and elimination rules for `Ex` in `Data.Fun.Extra`:
```idris
extractWitness : {ts : Vect n Type} -> {r : Rel ts} -> Ex ts r -> HVect ts
extractWitnessCorrect : {ts : Vect n Type} -> {r : Rel ts} -> (f : Ex ts r) -> uncurry {ts} r (extractWitness {r} f)
introduceWitness : {ts : Vect n Type} -> {r : Rel ts} -> (witness : HVect ts) -> uncurry {ts} r witness -> Ex ts r
```